File: node-engine/src/connection.rs
use serde::Serialize;
use serde_json::json;

use crate::errors::{Error, ErrorType};
use crate::node::NodeIndex;

#[derive(Debug, Clone)]
pub struct Connection {
    pub from_socket_type: SocketType,
    pub from_node: NodeIndex,
    pub to_socket_type: SocketType,
    pub to_node: NodeIndex,
}

impl Connection {
    pub fn serialize_to_json(&self) -> Result<serde_json::Value, Error> {
        Ok(json!([
            self.from_socket_type.serialize_to_json()?,
            self.from_node.index,
            self.to_socket_type.serialize_to_json()?,
            self.to_node.index
        ]))
    }
}

#[derive(Debug, Clone)]
pub struct InputSideConnection {
    pub from_socket_type: SocketType,
    pub from_node: NodeIndex,
    pub to_socket_type: SocketType,
}

#[derive(Debug, Clone)]
pub struct OutputSideConnection {
    pub from_socket_type: SocketType,
    pub to_node: NodeIndex,
    pub to_socket_type: SocketType,
}

#[derive(Debug, PartialEq, Clone, Serialize)]
pub enum SocketType {
    Stream(StreamSocketType),
    Midi(MidiSocketType),
    Value(ValueType),
    MethodCall(Vec<Parameter>),
}

impl SocketType {
    pub fn serialize_to_json(&self) -> Result<serde_json::Value, Error> {
        match serde_json::to_value(self) {
            Ok(result) => Ok(result),
            Err(error) => Err(Error::new(error.to_string(), ErrorType::ParserError)),
        }
    }
}

#[derive(Debug, PartialEq, Clone, Serialize)]
pub enum MidiSocketType {
    Default,
}

#[derive(Debug, PartialEq, Clone, Serialize)]
pub enum StreamSocketType {
    Audio,
    Gate,
    Detune,
    Dynamic(u64),
}

#[derive(Debug, PartialEq, Clone, Serialize)]
pub enum ValueType {
    Gain,
}

#[derive(Debug, PartialEq, Clone, Serialize)]
pub enum Parameter {
    Float(f32),
    Int(i32),
    Boolean(bool),
    String(String),
}

File: vpo-frontend-new/src/lib/node-engine/connection.ts
import { deepEqual } from "fast-equals";
import { Index } from "../ddgg/gen_vec";
import type { VertexIndex } from "../ddgg/graph";
import { type DiscriminatedUnion, match, matchOrElse } from "../util/discriminated-union";
import type { NodeConnection } from "./node_graph";


export type MidiSocketType = DiscriminatedUnion<"variant", {
    Default: {},
    Dynamic: { data: number }
}>;

export type StreamSocketType = DiscriminatedUnion<"variant", {
    Audio: {},
    Gate: {},
    Gain: {},
    Detune: {},
    Dynamic: { data: number },
}>;

export type ValueSocketType = DiscriminatedUnion<"variant", {
    Default: {},
    Gain: {},
    Frequency: {},
    Resonance: {},
    Gate: {},
    Attack: {},
    Decay: {},
    Sustain: {},
    Release: {},
    Speed: {},
    State: {},
    UiState: {},
    Dynamic: { data: number },
}>;

export type NodeRefSocketType = DiscriminatedUnion<"variant", {
    Button: {},
    Dynamic: { data: number },
}>;

export type Primitive = DiscriminatedUnion<"variant", {
    Float: { data: number },
    Int: { data: number },
    Boolean: { data: boolean },
    String: { data: string },
}>;

export type SocketType = DiscriminatedUnion<"variant", {
    Stream: { data: StreamSocketType },
    Midi: { data: MidiSocketType },
    Value: { data: ValueSocketType },
    NodeRef: { data: NodeRefSocketType },
}>;

export const SocketType = {
    toKey(socketType: SocketType) {
        return socketType.variant + "," + match(socketType, {
            Stream: ({ data: stream }) => matchOrElse(
                stream, {
                    Dynamic: ({ data }) => stream.variant + data,
                },  () => stream.variant
            ),
            Midi: ({ data: midi }) => matchOrElse(
                midi, {
                    Dynamic: ({ data }) => midi.variant + data,
                },  () => midi.variant
            ),
            Value: ({ data: value }) => matchOrElse(
                value, {
                    Dynamic: ({ data }) => value.variant + data,
                },  () => value.variant
            ),
            NodeRef:  ({ data: nodeRef }) => matchOrElse(
                nodeRef, {
                    Dynamic: ({ data }) => nodeRef.variant + data,
                },  () => nodeRef.variant
            ),
        });
    },
    areEqual(socketType1: SocketType, socketType2: SocketType): boolean {
        return deepEqual(socketType1, socketType2);
    },
}

export function socketToKey(socket: SocketType, direction: SocketDirection) {
    return SocketType.toKey(socket) + ":" + direction + match(socket, {
        Stream: ({ data: stream }) => matchOrElse(stream, {
            Dynamic: ({ data: uid }) => ":" + uid,
        },  () => "_"),
        Midi: ({ data: midi }) => matchOrElse(midi, {
            Dynamic: ({ data: uid }) => ":" + uid,
        },  () => "_"),
        Value: ({ data: value }) => matchOrElse(value, {
            Dynamic: ({ data: uid }) => ":" + uid,
        },  () => "_"),
        NodeRef: ({ data: nodeRef }) => matchOrElse(nodeRef, {
            Dynamic: ({ data: uid }) => ":" + uid,
        },  () => "_"),
    });
}

export enum SocketDirection {
    Input = 0,
    Output = 1
};

export interface Connection {
    fromNode: VertexIndex;
    toNode: VertexIndex;
    data: NodeConnection;
}

export const Connection = {
    getKey(connection: Connection): string {
        return SocketType.toKey(connection.data.fromSocketType) + ":" +
            Index.toKey(connection.fromNode) + "->" +
            SocketType.toKey(connection.data.toSocketType) + ":" +
            Index.toKey(connection.toNode);
    }
}

export interface InputSideConnection {
    fromSocketType: SocketType;
    fromNode: VertexIndex;
    toSocketType: SocketType;
}

export interface OutputSideConnection {
    fromSocketType: SocketType;
    toNode: VertexIndex;
    toSocketType: SocketType;
}
